cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll

```diff
@@ -0,0 +1,519 @@
+  }
+  or
+  HC_MemberFunctionCall(Function trg, HC qual, HC_Args args) {
+    mk_MemberFunctionCall(trg, qual, args, _)
+  }
+  or
+  // Any expression that is not handled by the cases above is
+  // given a unique number based on the expression itself.
+  HC_Unanalyzable(Expr e) { not analyzableExpr(e,_) }
+
+/** Used to implement hash-consing of argument lists  */
+private cached newtype HC_Args =
+  HC_EmptyArgs(Function fcn) {
+    any()
+  }
+  or HC_ArgCons(Function fcn, HC hc, int i, HC_Args list) {
+    mk_ArgCons(fcn, hc, i, list, _)
+  }
+/**
+ * HC is the hash-cons of an expression. The relationship between `Expr`
+ * and `HC` is many-to-one: every `Expr` has exactly one `HC`, but multiple
+ * expressions can have the same `HC`. If two expressions have the same
+ * `HC`, it means that they are structurally equal. The `HC` is an opaque
+ * value. The only use for the `HC` of an expression is to find other
+ * expressions that are structurally equal to it. Use the predicate
+ * `hashCons` to get the `HC` for an `Expr`.
+ *
+ * Note: `HC` has `toString` and `getLocation` methods, so that it can be
+ * displayed in a results list. These work by picking an arbitrary
+ * expression with this `HC` and using its `toString` and `getLocation`
+ * methods.
+ */
+class HC extends HCBase {
+  HC() { this instanceof HCBase }
+
+  /** Gets an expression that has this HC. */
+  Expr getAnExpr() {
+    this = hashCons(result)
+  }
+
+  /** Gets the kind of the HC. This can be useful for debugging. */
+  string getKind() {
+    if this instanceof HC_IntLiteral then result = "IntLiteral" else
+    if this instanceof HC_FloatLiteral then result = "FloatLiteral" else
+    if this instanceof HC_StringLiteral then result = "StringLiteral" else
+    if this instanceof HC_Variable then result = "Variable" else
+    if this instanceof HC_FieldAccess then result = "FieldAccess" else
+    if this instanceof HC_Deref then result = "Deref" else
+    if this instanceof HC_ThisExpr then result = "ThisExpr" else
+    if this instanceof HC_Conversion then result = "Conversion" else
+    if this instanceof HC_BinaryOp then result = "BinaryOp" else
+    if this instanceof HC_UnaryOp then result = "UnaryOp" else
+    if this instanceof HC_ArrayAccess then result = "ArrayAccess" else
+    if this instanceof HC_Unanalyzable then result = "Unanalyzable" else
+    if this instanceof HC_NonmemberFunctionCall then result = "NonmemberFunctionCall" else
+    if this instanceof HC_MemberFunctionCall then result = "MemberFunctionCall" else
+    result = "error"
+  }
+
+  /**
+   * Gets an example of an expression with this HC.
+   * This is useful for things like implementing toString().
+   */
+  private Expr exampleExpr() {
+    // Pick the expression with the minimum source location string. This is
+    // just an arbitrary way to pick an expression with this `HC`.
+    result =
+       min(Expr e
+       | this = hashCons(e)
+       | e order by e.getLocation().toString())
```

Comment:
There should only be 1 Location for any physical location of expressions. There may be another one for the location of statements, and another for the location of other things.

Reply:
The results there are very strange; it gives one location for the literal and two for the variable access and the addition. I think that's extractor weirdness, but it might be something in the library

Reply:
to be clear, that seems to be one location for each variable access in the macro expansion, with `loc1 != loc2`, but `loc1.subsumes(loc2)` and `loc2.subsumes(loc1)`

Reply:
added a test; I'm still not sure what's going on here. @nickrolfe @ian-semmle can you comment on what's happening in this case?

Reply:
It does look like the accesses to `y` are duplicated, but somehow there's only one row in the `.expected` file for the `HashCons` object even though I'd expect it to have two locations. So it might be strange, but I suppose it's not a problem.